File: engine/src/state/block_map.rs
use crate::provenance::attention::ranges_overlap;
use crate::types::{CodeBlock, Origin};

/// Tracks origin/attention for the blocks of a single file, keeping line
/// ranges in sync as the document is edited. Port of the TS `blockMap.ts`.
#[derive(Default)]
pub struct FileBlocks {
    /// Kept for parity with the TS engine / future per-file reporting.
    #[allow(dead_code)]
    pub fs_path: String,
    pub blocks: Vec<CodeBlock>,
    next_id: u32,
}

impl FileBlocks {
    pub fn new(fs_path: String) -> Self {
        FileBlocks {
            fs_path,
            blocks: Vec::new(),
            next_id: 1,
        }
    }

    pub fn add_block(
        &mut self,
        start_line: u32,
        end_line: u32,
        origin: Origin,
        char_count: u32,
        now_ms: f64,
    ) {
        // Skip if fully inside an existing same-origin block.
        for b in &self.blocks {
            if b.origin == origin && start_line >= b.start_line && end_line <= b.end_line {
                return;
            }
        }
        let id = self.next_id;
        self.next_id += 1;
        self.blocks.push(CodeBlock {
            id,
            start_line,
            end_line,
            origin,
            created_at_ms: now_ms,
            read: false,
            edited: false,
            ran: false,
            char_count,
        });
    }

    /// Apply a document edit, shifting blocks below and resizing overlaps.
    pub fn apply_shift(&mut self, change_start: u32, change_end: u32, line_delta: i64) {
        if line_delta == 0 {
            return;
        }
        for b in &mut self.blocks {
            if b.start_line > change_end {
                b.start_line = shift(b.start_line, line_delta);
                b.end_line = shift(b.end_line, line_delta);
            } else if ranges_overlap(change_start, change_end, b.start_line, b.end_line) {
                let new_end = shift(b.end_line, line_delta);
                b.end_line = new_end.max(b.start_line);
            }
        }
    }

    /// Explicitly mark the block containing this line as reviewed (the only clear).
    pub fn mark_reviewed_at(&mut self, line: u32) {
        for b in &mut self.blocks {
            if !b.read && line >= b.start_line && line <= b.end_line {
                b.read = true;
            }
        }
    }

    pub fn mark_edited(&mut self, line: u32) {
        for b in &mut self.blocks {
            if line >= b.start_line && line <= b.end_line {
                b.edited = true;
            }
        }
    }

    pub fn mark_all_read(&mut self) {
        for b in &mut self.blocks {
            b.read = true;
        }
    }
}

fn shift(value: u32, delta: i64) -> u32 {
    let result = value as i64 + delta;
    if result < 0 {
        0
    } else {
        result as u32
    }
}

File: extension/src/engine/blockMap.ts
import { CodeBlock, Origin } from "./types";

function overlaps(aStart: number, aEnd: number, bStart: number, bEnd: number): boolean {
  return aStart <= bEnd && bStart <= aEnd;
}

/**
 * Tracks origin/attention for the blocks of a single file.
 *
 * The line ranges are kept in sync with the live document: every edit shifts
 * the blocks below it (insert/delete) so a stored block always points at the
 * same logical code even as the file grows above it. This is the TS port of
 * the Rust `block_map.rs`.
 */
export class FileBlocks {
  blocks: CodeBlock[] = [];
  private nextId = 1;

  constructor(public readonly fsPath: string) {}

  addBlock(startLine: number, endLine: number, origin: Origin, charCount: number): void {
    // If this insertion sits fully inside an existing unread external block of
    // the same origin, treat it as part of that block instead of duplicating.
    for (const b of this.blocks) {
      if (b.origin === origin && startLine >= b.startLine && endLine <= b.endLine) {
        return;
      }
    }
    this.blocks.push({
      id: this.nextId++,
      startLine,
      endLine,
      origin,
      createdAt: Date.now(),
      read: false,
      edited: false,
      charCount,
    });
  }

  /**
   * Apply a document edit to keep block ranges accurate.
   * @param changeStartLine first line touched by the edit
   * @param changeEndLine   last line touched (before edit)
   * @param lineDelta       net change in line count (added - removed)
   */
  applyShift(changeStartLine: number, changeEndLine: number, lineDelta: number): void {
    if (lineDelta === 0) {
      return;
    }
    for (const b of this.blocks) {
      if (b.startLine > changeEndLine) {
        // Block entirely below the edit: shift it.
        b.startLine += lineDelta;
        b.endLine += lineDelta;
      } else if (overlaps(changeStartLine, changeEndLine, b.startLine, b.endLine)) {
        // Edit overlaps the block: grow/shrink its end.
        b.endLine = Math.max(b.startLine, b.endLine + lineDelta);
      }
    }
  }

  /** Explicitly mark the block containing this line as reviewed (the only clear). */
  markReviewedAt(line: number): boolean {
    let changed = false;
    for (const b of this.blocks) {
      if (!b.read && line >= b.startLine && line <= b.endLine) {
        b.read = true;
        changed = true;
      }
    }
    return changed;
  }

  /** A manual edit landed on this line — the user actively changed it. */
  markEdited(line: number): void {
    for (const b of this.blocks) {
      if (line >= b.startLine && line <= b.endLine) {
        b.edited = true;
      }
    }
  }

  /** Force every block to reviewed (used by "Mark file as reviewed"). */
  markAllRead(): void {
    for (const b of this.blocks) {
      b.read = true;
    }
  }
}
